pkg/virt-launcher/virtwrap/network/podinterface_test.go

```diff
@@ -92,9 +95,12 @@ var _ = Describe("Pod Network", func() {
 		testMac := "12:34:56:78:9A:BC"
 		updateTestMac := "AF:B3:1F:78:2A:CA"
 		mtu = 1410
-		dummySwap = &netlink.Dummy{LinkAttrs: netlink.LinkAttrs{Name: primaryPodInterfaceName}}
-		dummy = &netlink.Dummy{LinkAttrs: netlink.LinkAttrs{Index: 1, MTU: mtu}}
+		originalPodInterfaceName = primaryPodInterfaceName
 		newPodInterfaceName = fmt.Sprintf("%s-nic", primaryPodInterfaceName)
+		dummySwap = &netlink.Dummy{LinkAttrs: netlink.LinkAttrs{Name: originalPodInterfaceName}}
+		primaryPodInterfaceBeforeNameChange = &netlink.GenericLink{LinkAttrs: netlink.LinkAttrs{Name: originalPodInterfaceName}}
```

Comment:
Although I suggested both of the options, I have to admit that I prefer removing `if b.podNicLink.Type() != "dummy"` fromthe `switchPodInterfaceWithDummy` method. It is redundant there.

Reply:
Testing fails on `should define a new VIF bind to a bridge`.  I like keeping it just to ensure we're very explicit about this function.  Have that `if` check and using `addrReplace` make sure we're only fixing precisely the links that need fixing.

Reply:
If the test fails, we should understand why and fix it. Can you please attach the error you get so I will try to help?
`switchPodInterfaceWithDummy` is an internal method. We know it is called only when `b.podNicLink` is not yet the dummy one.
I believe that having this `if` in the code may confuse future developers that may understand there is a situation `b.podNicLink` can be the dummy one in this stage. Which AFAIU, is not possible.

The fact that removing this `if` statement caused errors in the unit tests is really good. That means that now the tests actually cover the new code (`switchPodInterfaceWithDummy`) and previously it just was skipped.

Reply:
```
==================== Test output for //pkg/virt-launcher/virtwrap/network:go_default_test:
Running Suite: Network Suite
============================
Random Seed: 1612179467
Will run 40 of 40 specs

• Failure [0.004 seconds]
Pod Network
pkg/virt-launcher/virtwrap/network/podinterface_test.go:45
  on successful setup
  pkg/virt-launcher/virtwrap/network/podinterface_test.go:307
    should define a new VIF bind to a bridge [It]
    pkg/virt-launcher/virtwrap/network/podinterface_test.go:308

    Unexpected call to *network.MockNetworkHandler.LinkSetName([0xc000012480 eth0-nic]) at pkg/virt-launcher/virtwrap/network/generated_mock_common.go:121 because:
    Expected call at pkg/virt-launcher/virtwrap/network/generated_mock_common.go:127 doesn't match the argument at index 0.
    Got: &{{1 1410 0   0 0 0 0 <nil>  <nil> 0 <nil>  <nil> unknown 0 0 0 0 0 [] 0 <nil>}}
    Want: is equal to &{{0 0 0 eth0  0 0 0 0 <nil>  <nil> 0 <nil>  <nil> unknown 0 0 0 0 0 [] 0 <nil>} }
```

My concern around the `if` block Alona is that we're changing the interface name and creating a new one every time `switchPodInterfaceWithDummy` is called.  Without the `if` statement, that function is not idempotent, and I'm not certain can handle multiple restarts where each restart leads to the pod nic being constantly appending like: `eth0-nic-nic-nic-nic ....`.  This puts a lot of weight on the `b.vif.IPAMDisabled` if statement or the `switchPodInterfaceWithDummy` code path not being called.  Maybe you can help fill in the context here that I might be missing.

Reply:
I believe that if you will remove `mockNetwork.EXPECT().LinkByName(primaryPodInterfaceName).Return(dummy, nil).Times(2)` you won't get this error.

The code in `PlugPhase1`->`preparePodInterface` is not idempotent. It is designed to be called only once per virt-launcher pod.
For example, the mac address is removed from the podInterface and the pod interface gets a new random mac (the original mac is stored on the virtInterface to be passed to libvirt/the vm).
If `PlugPhase1`->`preparePodInterface` will be invoked more than once than the vm will get the wrong mac (the random one we put on the podInterface in the previous execution).

Reply:
Thanks for context.

5 errors after I remove that line.  All the same error but in different places.

```
    Unexpected call to *network.MockNetworkHandler.AddrList([0xc0005c4360 0]) at pkg/virt-launcher/virtwrap/network/generated_mock_common.go:49 because:
    Expected call at pkg/virt-launcher/virtwrap/network/generated_mock_common.go:56 doesn't match the argument at index 0.
    Got: &{{0 0 0 eth0  0 0 0 0 <nil>  <nil> 0 <nil>  <nil> unknown 0 0 0 0 0 [] 0 <nil>} }
    Want: is equal to &{{1 1410 0   0 0 0 0 <nil>  <nil> 0 <nil>  <nil> unknown 0 0 0 0 0 [] 0 <nil>}}
    Expected call at pkg/virt-launcher/virtwrap/network/generated_mock_common.go:56 doesn't match the argument at index 0.
    Got: &{{0 0 0 eth0  0 0 0 0 <nil>  <nil> 0 <nil>  <nil> unknown 0 0 0 0 0 [] 0 <nil>} }
    Want: is equal to &{{1 1410 0   0 0 0 0 <nil>  <nil> 0 <nil>  <nil> unknown 0 0 0 0 0 [] 0 <nil>}}
    Expected call at pkg/virt-launcher/virtwrap/network/generated_mock_common.go:56 doesn't match the argument at index 0.
    Got: &{{0 0 0 eth0  0 0 0 0 <nil>  <nil> 0 <nil>  <nil> unknown 0 0 0 0 0 [] 0 <nil>} }
    Want: is equal to &{{1 1410 0   0 0 0 0 <nil>  <nil> 0 <nil>  <nil> unknown 0 0 0 0 0 [] 0 <nil>}}
```
1) Masquerade Plug
  -   should define a new VIF bind to a bridge and create a default nat rule using iptables [It]
      pkg/virt-launcher/virtwrap/network/podinterface_test.go:433
  -   should define a new VIF bind to a bridge and create a specific nat rule using iptables [It]
      pkg/virt-launcher/virtwrap/network/podinterface_test.go:448
  -  should define a new VIF bind to a bridge and create a default nat rule using nftables [It]                                                                                                                                                      
      pkg/virt-launcher/virtwrap/network/podinterface_test.go:485 
  -   should define a new VIF bind to a bridge and create a specific nat rule using nftables [It]
      pkg/virt-launcher/virtwrap/network/podinterface_test.go:499
2) should define a new VIF bind to a bridge [It]                                                                                                                                                                                                pkg/virt-launcher/virtwrap/network/podinterface_test.go:307

Reply:
You have to replace all the places that are using `dummy` for the bridge binding test to use `primaryPodInterfaceBeforeNameChange\primaryPodInterfaceAfterNameChange`/
In the specific error you pasted - [AddrList](https://github.com/kubevirt/kubevirt/blob/3a472736142f8449868ddbbf5d75bd70de76b5da/pkg/virt-launcher/virtwrap/network/podinterface_test.go#L183) has to be changed.
I would change the `dummy` var name to `dummyMasquerade` so it will be clear no bridge binding test should use it.